src/conn/conn_prefetch.c

```diff
@@ -0,0 +1,176 @@
+/*-
+ * Copyright (c) 2014-present MongoDB, Inc.
+ * Copyright (c) 2008-2014 WiredTiger, Inc.
+ *	All rights reserved.
+ *
+ * See the file LICENSE for redistribution information.
+ */
+
+#include "wt_internal.h"
+
+/*
+ * __wt_prefetch_create --
+ *     Start the pre-fetch server.
+ */
+int
+__wt_prefetch_create(WT_SESSION_IMPL *session, const char *cfg[])
+{
+    WT_CONFIG_ITEM cval;
+    WT_CONNECTION_IMPL *conn;
+    uint32_t session_flags;
+
+    conn = S2C(session);
+
+    /*
+     * This might have already been parsed and set during connection configuration, but do it here
+     * as well, in preparation for the functionality being runtime configurable.
+     */
+    WT_RET(__wt_config_gets(session, cfg, "prefetch", &cval));
+    conn->prefetch_auto_on = cval.val != 0;
+
+    /*
+     * Pre-fetch functionality isn't runtime configurable, so don't bother starting utility threads
+     * if it isn't enabled.
+     */
+    if (!conn->prefetch_auto_on)
+        return (0);
+
+    F_SET(conn, WT_CONN_PREFETCH_RUN);
+
+    session_flags = WT_THREAD_CAN_WAIT | WT_THREAD_PANIC_FAIL;
+    WT_RET(__wt_thread_group_create(session, &conn->prefetch_threads, "prefetch-server", 8, 8,
+      session_flags, __wt_prefetch_thread_chk, __wt_prefetch_thread_run, NULL));
+
+    return (0);
+}
+
+/*
+ * __wt_prefetch_thread_chk --
+ *     Check to decide if the pre-fetch thread should continue running.
+ */
+bool
+__wt_prefetch_thread_chk(WT_SESSION_IMPL *session)
+{
+    return (F_ISSET(S2C(session), WT_CONN_PREFETCH_RUN));
+}
+
+/*
+ * __wt_prefetch_thread_run --
+ *     Entry function for a prefetch thread. This is called repeatedly from the thread group code so
+ *     it does not need to loop itself.
+ */
+int
+__wt_prefetch_thread_run(WT_SESSION_IMPL *session, WT_THREAD *thread)
+{
+    WT_CONNECTION_IMPL *conn;
+    WT_DECL_ITEM(tmp);
+    WT_DECL_RET;
+    WT_PREFETCH *pf;
+    bool locked;
+
+    WT_UNUSED(thread);
+
+    WT_ASSERT(session, session->id != 0);
+
+    conn = S2C(session);
+    locked = false;
+    WT_RET(__wt_scr_alloc(session, 0, &tmp));
+
+    while (F_ISSET(conn, WT_CONN_PREFETCH_RUN)) {
```

Comment:
I may be wrong but I think we were doing a sleep because we don't have any mechanism to be notified when candidate pages for prefetching get added to the queue. We plan to do another review and make changes to this area in WT-11691.

Reply:
That works for me.